site/app/libraries/database/PostgresqlDatabaseQueries.php

```diff
@@ -1277,5 +1278,200 @@ function($elem) { return intval($elem); },
+                  array_agg(gc_max_value) AS array_max_value,
+                  array_agg(gc_upper_clamp) AS array_upper_clamp,
+                  array_agg(gc_is_text) AS array_text,
+                  array_agg(gc_is_peer) AS array_peer,
+                  array_agg(gc_order) AS array_order,
+                  array_agg(gc_page) AS array_page,
+                  array_agg(array_to_json(gcm.array_id)) AS array_mark_id,
+                  array_agg(array_to_json(gcm.array_points)) AS array_mark_points,
+                  array_agg(array_to_json(gcm.array_title)) AS array_mark_title,
+                  array_agg(array_to_json(gcm.array_publish)) AS array_mark_publish,
+                  array_agg(array_to_json(gcm.array_order)) AS array_mark_order
+                FROM gradeable_component gc
+                LEFT JOIN (
+                  SELECT
+                    gc_id AS gcm_gc_id,
+                    array_agg(gcm_id) AS array_id,
+                    array_agg(gcm_points) AS array_points,
+                    array_agg(gcm_note) AS array_title,
+                    array_agg(gcm_publish) AS array_publish,
+                    array_agg(gcm_order) AS array_order
+                    FROM gradeable_component_mark
+                  GROUP BY gcm_gc_id
+                ) AS gcm ON gcm.gcm_gc_id=gc.gc_id
+                GROUP BY g_id
+              ) AS gc ON gc.gc_g_id=g.g_id
+             $selector
+             ORDER BY g.g_id";
+
+        $gradeable_constructor = function($row) {
+            if (!isset($row['eg_g_id']) && $row['type'] === GradeableType::ELECTRONIC_FILE) {
+                throw new DatabaseException("Electronic gradeable didn't have an entry in the electronic_gradeable table!");
+            }
+
+            // Finally, create the gradeable
+            $gradeable = new \app\models\gradeable\Gradeable($this->core, $row);
+
+            // Construct the components
+            $component_properties = [
+                'id',
+                'title',
+                'ta_comment',
+                'student_comment',
+                'lower_clamp',
+                'default',
+                'max_value',
+                'upper_clamp',
+                'text',
+                'peer',
+                'order',
+                'page'
+            ];
+            $mark_properties = [
+                'id',
+                'points',
+                'title',
+                'publish',
+                'order'
+            ];
+            $component_mark_properties = array_map(function ($value) {
+                return 'mark_' . $value;
+            }, $mark_properties);
+
+            // Unpack the component data
+            $unpacked_component_data = [];
+            foreach (array_merge($component_properties, $component_mark_properties) as $property) {
+                $unpacked_component_data[$property] = $this->course_db->fromDatabaseToPHPArray($row['array_' . $property]);
```

Comment:
I'd also recommend json_decode for these

Reply:
I switched to using the json method.  I don't have a precise way to test what the performance difference is between the two though.